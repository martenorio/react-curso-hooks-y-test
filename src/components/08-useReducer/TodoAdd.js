import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {

    const [{ description },handleInputChange, reset] = useForm({
        description:""
    })

    const handleSubmit = (e) => {
        e.preventDefault();
        if(description.trim().length <= 1) return;
        const newTodo = {
            id:new Date().getTime(),
            desc:description,
            done:false
        };
        handleAddTodo(newTodo);
        reset();
    }

    return (
        <>
            <h4>Agregar Todo</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input className='form-control'
                    onChange={handleInputChange}
                    value={description}
                    type="text" name='description'
                    placeholder='Aprender...'
                    autoComplete='off' />
                <div className="d-grid gap-2">
                    <button className="btn btn-outline-primary mt-2 btn-block">Agregar</button>
                </div>
            </form>
        </>
    )
}
