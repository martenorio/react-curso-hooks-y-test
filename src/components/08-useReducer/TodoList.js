import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({todos,handleDelete,handleToggle}) => {
  return (
    <ul className='list-group list-group-flush'>
        { todos.map(
            (t,i) => <TodoListItem key={ i } todo={ t } index={ i } handleDelete={ handleDelete } handleToggle={ handleToggle } />
            )
        }
    </ul>
  )
}
