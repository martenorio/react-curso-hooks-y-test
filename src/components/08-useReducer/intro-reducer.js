const initialState = [{
    id:1,
    todo:'Compras de supermercado',
    done: false
}];

const todoReducer = (state = initialState, action) => {
    if(action?.type === "agregar"){
         return [...state, action.payload]
    }
    return state;
}

let todos = todoReducer();

// No utilizar el push en react por que muta el valor de los estado
// y puede causar un error o no hacer nada.(se pudiera utilizar en useRef)

const newTodo = {
    id:2,
    todo:'Pagar recibos de luz',
    done: false
}

const agregarTodoAction = {
    type: 'agregar',
    payload:newTodo
}

todos = todoReducer(todos, agregarTodoAction);

console.log(todos);