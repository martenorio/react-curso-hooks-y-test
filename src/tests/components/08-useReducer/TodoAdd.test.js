import { shallow } from 'enzyme';
import { TodoAdd } from '../../../components/08-useReducer/TodoAdd';

describe('TodoAdd test', () => {
    const handleAddTodo = jest.fn();
    const wrapper = shallow( <TodoAdd handleAddTodo={ handleAddTodo } /> )
    
    test('should display correctly', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('shouldnÂ´t call handleAddTodo', () => {
        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({preventDefault: ()=>{}});
        expect(handleAddTodo).toHaveBeenCalledTimes(0);
    });
    test('should call handleAddTodo funtion', () => {
        const value = "nombre";
        wrapper.find('input').simulate('change', { target: { value, name:'description' } } );
        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({preventDefault: ()=>{}});
        expect(handleAddTodo).toHaveBeenCalledTimes(1);
        expect(handleAddTodo).toHaveBeenCalledWith( expect.any(Object) ) // {}
        expect(handleAddTodo).toHaveBeenCalledWith( {
            id: expect.any(Number),
            desc:value,
            done:false
        } )
        expect( wrapper.find('input').prop('value') ).toBe('')
    });
})