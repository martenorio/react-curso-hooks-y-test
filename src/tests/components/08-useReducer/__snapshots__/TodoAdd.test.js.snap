// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoAdd test should display correctly 1`] = `
<Fragment>
  <h4>
    Agregar Todo
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprender..."
      type="text"
      value=""
    />
    <div
      className="d-grid gap-2"
    >
      <button
        className="btn btn-outline-primary mt-2 btn-block"
      >
        Agregar
      </button>
    </div>
  </form>
</Fragment>
`;
